Improve and refactor cellseqs_v2 in Nim standard library

jordan-gillard opened the pull request:
**Description:**

This pull request introduces a set of improvements and refactoring to the `cellseqs_v2` module in the standard library. The changes aim to enhance code readability, maintainability, and memory efficiency.

**Key changes:**

1. Extract the resizing logic from the `add` procedure into a separate `resize` procedure. This change improves code readability and separates concerns, making it easier to understand the purpose and function of each procedure.
2. Implement `realloc` for non-threaded cases in the `resize` procedure, which reduces memory operations and can lead to performance improvements.
3. Change the `add` procedure's signature to use `,` instead of `;` between parameters for consistency with other code.

These changes should not alter the core functionality of the `cellseqs_v2` module but will make it more maintainable and easier to understand for future contributors.

jordan-gillard committed:
Refactor and optimize cellseqs_v2 in Nim standard library

* Extract resizing logic into a separate 'resize' procedure for better readability and separation of concerns
* Implement realloc for non-threaded cases to improve memory operations efficiency
* Use ',' instead of ';' between parameters in 'add' procedure for consistency with other Nim code

juancarlospaco commented on lib/system/cellseqs_v2.nim:
```diff
@@ -16,20 +16,21 @@ type
-proc add[T](s: var CellSeq[T], c: T; t: PNimTypeV2) {.inline.} =
+proc resize[T](s: var CellSeq[T]) =
+  s.cap = s.cap * 3 div 2
+  var newSize = s.cap * sizeof(CellTuple[T])
```
`newSize` is only used once and its tiny, I would just inline it directly IMHO.

jordan-gillard replied:
Its used on line 25 and 29 🥲 

Inlining the variable changes line 25 to 
```
s.d = cast[CellArray[T]](allocShared(s.cap * sizeof(CellTuple[T])))
``` 
and line 29 to 
```
s.d = cast[CellArray[T]](realloc(s.d, s.cap * sizeof(CellTuple[T])))
```
I'm a fan of the readability improvement. 

Is this a dealbreaker?

beef331 replied:
If we're bikeshedding or code golfing it, a template is more concise than manually in-lining. still self documenting and less redundant.

jordan-gillard replied:
```nim
template calcNewSize(s: CellSeq): int =
  s.cap * sizeof(CellTuple[s.type])

proc resize[T](s: var CellSeq[T]) =
  s.cap = s.cap * 3 div 2

  when compileOption("threads"):
    let oldData = s.d
    s.d = cast[CellArray[T]](allocShared(calcNewSize(s)))
    copyMem(s.d, oldData, s.len * sizeof(CellTuple[T]))
    deallocShared(oldData)
  else:
    s.d = cast[CellArray[T]](realloc(s.d, calcNewSize(s)))
```
I think using a variable is more readable than a template, but the difference in readability likely comes down to developer preference. ~I'd be fine with this change.~ Edit: I change my mind

I'm not sure if either of you want me to refactor this code - but now there's 3 differing opinions on how to represent the expression `s.cap * sizeof(CellTuple[T])`.

beef331 replied:
Like i said *If we're bike shedding* I could not careless. :smile:

jordan-gillard replied:
@beef331 @juancarlospaco can I get an approval or rereview then? This is the most confusing review process I have ever participated in.

beef331 replied:
Hey it says "collaborator" not "contributor" next to my name. :stuck_out_tongue: 
The code seems fine to me, but it does not matter what I say.

juancarlospaco replied:
> Its used on line 25 and 29 smiling_face_with_tear

If you have `when` branches then only 1 will actually exist whatsoever.

`resize` should be a `template` IMHO.

jordan-gillard replied:
> If you have `when` branches then only 1 will actually exist whatsoever.
> 
> `resize` should be a `template` IMHO.

Templates are useful when you want to generate code at compile time by substituting patterns. They are helpful for avoiding code duplication, improving performance by removing function call overhead, and customizing behavior through parameters.

In the case of the `resize` function, the expression `s.cap * sizeof(CellTuple[T])` is only used once, and it doesn't contain any complex logic that could benefit from being extracted into a separate template. Therefore creating a template for this single expression doesn't make sense since there isn't a clear benefit from doing it.

metagn replied:
Just variable seems fine here?

Araq commented on lib/system/cellseqs_v2.nim:
```diff
@@ -16,20 +16,21 @@ type
+
+  when compileOption("threads"):
+    let oldData = s.d
+    s.d = cast[CellArray[T]](allocShared(newSize))
```
We have `reallocShared` too.

jordan-gillard replied:
Oooo I don't know why I didn't see that before! The code looks _way_ better now. Addressed here: 9fdb2f06f1e78b2af77479780f432c2c6ab07b4f

jordan-gillard committed:
Respond to Araq's feedback: Refactor resize function to use reallocShared

This commit replaces the usage of allocShared and deallocShared with reallocShared to optimize memory allocation and deallocation while resizing the CellSeq.

jordan-gillard committed:
Merge branch 'devel' into patch-1
